File: semester2/5_Fuzzy-Software-Technologies/aics-fuzzy-software-technologies/ticket-backend/src/main/java/aics/domain/fuzzy/models/FuzzyVariableRating.java
package aics.domain.fuzzy.models;

import aics.domain.fuzzy.constants.FuzzyVariableRatingFields;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

import java.io.Serializable;
import java.util.HashMap;

@Data
@Accessors(chain = true)
@AllArgsConstructor
@NoArgsConstructor
public class FuzzyVariableRating implements Serializable, FuzzyVariableI {
    private FuzzyVariableDistributionPart varBad;
    private FuzzyVariableDistributionPart varAverage;
    private FuzzyVariableDistributionPart varGood;
    private FuzzyVariableDistributionPart varVeryGood;

    @Override
    public HashMap<String, FuzzyVariableDistributionPart> getFuzzyVariableMap() {
        HashMap<String, FuzzyVariableDistributionPart> fuzzyVariableMap = new HashMap<>();
        fuzzyVariableMap.put(FuzzyVariableRatingFields.BAD.name(), this.varBad);
        fuzzyVariableMap.put(FuzzyVariableRatingFields.AVERAGE.name(), this.varAverage);
        fuzzyVariableMap.put(FuzzyVariableRatingFields.GOOD.name(), this.varGood);
        fuzzyVariableMap.put(FuzzyVariableRatingFields.VERY_GOOD.name(), this.varVeryGood);
        return fuzzyVariableMap;
    }

    @Override
    public FuzzyVariableDistributionPart find1stPart() {
        return this.varBad;
    }

    @Override
    public FuzzyVariableDistributionPart find2ndPart() {
        return this.varAverage;
    }

    @Override
    public FuzzyVariableDistributionPart find3rdPart() {
        return this.varGood;
    }

    @Override
    public FuzzyVariableDistributionPart find4thPart() {
        return this.varVeryGood;
    }
}

File: semester2/5_Fuzzy-Software-Technologies/aics-fuzzy-software-technologies/ticket-frontend/src/modules/fuzzy/models/fuzzy-variable-rating.ts
import { FuzzyVariableDistributionPart, FuzzyVariableDistributionPartUtils, FuzzyVariableI } from './fuzzy-variable-distribution';

export enum FuzzyVariableRatingFields {
    BAD = 'BAD',
    AVERAGE = 'AVERAGE',
    GOOD = 'GOOD',
    VERY_GOOD = 'VERY_GOOD',
}

export class FuzzyVariableRating implements FuzzyVariableI {
    varBad: FuzzyVariableDistributionPart;
    varAverage: FuzzyVariableDistributionPart;
    varGood: FuzzyVariableDistributionPart;
    varVeryGood: FuzzyVariableDistributionPart;

    constructor(obj: {
        varBad: FuzzyVariableDistributionPart,
        varAverage: FuzzyVariableDistributionPart,
        varGood: FuzzyVariableDistributionPart,
        varVeryGood: FuzzyVariableDistributionPart
    }) {
        this.varBad = obj.varBad;
        this.varAverage = obj.varAverage;
        this.varGood = obj.varGood;
        this.varVeryGood = obj.varVeryGood;
    }


    set1stPart(fuzzyVariableDistributionPart: FuzzyVariableDistributionPart) {
        this.varBad = fuzzyVariableDistributionPart;
    }
    set2ndPart(fuzzyVariableDistributionPart: FuzzyVariableDistributionPart) {
        this.varAverage = fuzzyVariableDistributionPart;
    }
    set3rdPart(fuzzyVariableDistributionPart: FuzzyVariableDistributionPart) {
        this.varGood = fuzzyVariableDistributionPart;
    }
    set4thPart(fuzzyVariableDistributionPart: FuzzyVariableDistributionPart) {
        this.varVeryGood = fuzzyVariableDistributionPart;
    }

    getName(): string {
        return "RATING";
    }

    get1stPart() {
        return this.varBad;
    }

    get2ndPart() {
        return this.varAverage;
    }

    get3rdPart() {
        return this.varGood;
    }

    get4thPart() {
        return this.varVeryGood;
    }

    static fromObj(obj: any): FuzzyVariableRating {
        return new FuzzyVariableRating({
            varBad: FuzzyVariableDistributionPartUtils.fuzzyVariableDistributionPartFromObj(obj.varBad),
            varAverage: FuzzyVariableDistributionPartUtils.fuzzyVariableDistributionPartFromObj(obj.varAverage),
            varGood: FuzzyVariableDistributionPartUtils.fuzzyVariableDistributionPartFromObj(obj.varGood),
            varVeryGood: FuzzyVariableDistributionPartUtils.fuzzyVariableDistributionPartFromObj(obj.varVeryGood)
        })
    }

    deepClone(): FuzzyVariableRating {
        return new FuzzyVariableRating({
            varBad: this.varBad.deepClone(),
            varAverage: this.varAverage.deepClone(),
            varGood: this.varGood.deepClone(),
            varVeryGood: this.varVeryGood.deepClone(),
        })
    }

    getFuzzyVariableMap(): Record<string, FuzzyVariableDistributionPart> {
        return {
            [FuzzyVariableRatingFields.BAD]: this.varBad,
            [FuzzyVariableRatingFields.AVERAGE]: this.varAverage,
            [FuzzyVariableRatingFields.GOOD]: this.varGood,
            [FuzzyVariableRatingFields.VERY_GOOD]: this.varVeryGood,
        }
    }

    getFuzzyVariableColorsMap(): Record<string, string> {
        return {
            [FuzzyVariableRatingFields.BAD]: 'red',
            [FuzzyVariableRatingFields.AVERAGE]: 'orange',
            [FuzzyVariableRatingFields.GOOD]: 'yellow',
            [FuzzyVariableRatingFields.VERY_GOOD]: 'green',
        }
    }
}
